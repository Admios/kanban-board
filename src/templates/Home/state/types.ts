export interface Todo {
  id: string;
  columnId: string;
  text: string;
  position: number;
  owner: string;
}

export interface Column {
  id: string;
  name: string;
  position: number;
  color: string;
  owner: string;
}
