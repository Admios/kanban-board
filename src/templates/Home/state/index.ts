import { ColumnDTO as DbColumn } from "@/model/Column/ColumnDTO";
import { TaskDTO as DBTodo } from "@/model/Task/TaskDTO";
import { UserDTO } from "@/model/User/UserDTO";
import { Immutable, produce } from "immer";
import { v4 as uuid } from "uuid";
import { StateCreator, create } from "zustand";
import { Column, Todo } from "./types";

export type { Column, Todo } from "./types";

type HomeState = Immutable<{
  todos: Record<string, Todo[]>;
  columns: Record<string, Column>;
  user?: UserDTO;
}>;

interface HomeActions {
  initialize(
    initialTodos: DBTodo[],
    initialColumns: DbColumn[],
    initialUser?: UserDTO,
  ): void;
  addTodo(newTodo: Omit<Todo, "id">): Todo;
  moveTodo(
    newTodo: Todo,
    fromColumnId: string,
    toColumnId: string,
    position: number,
  ): void;
  addColumn(newColumn: Omit<Column, "id" | "position">): Column;
}

const stateCreator: StateCreator<HomeState & HomeActions> = (set, get) => ({
  todos: {},
  columns: {},

  initialize(initialTodos, initialColumns, initialUser) {
    const columnMap: Record<string, Column> = {};
    const todosMap: Record<string, Todo[]> = {};
    const backendIdToIdMap = new Map<string, string>();

    initialColumns.forEach((backendColumn) => {
      const result: Column = {
        ...backendColumn,
        id: uuid(),
        backendId: backendColumn.id,
      };
      columnMap[result.id] = result;
      if (result.backendId) {
        backendIdToIdMap.set(result.backendId, result.id);
      }
    });

    initialTodos.forEach((backendTodo) => {
      const columnId =
        backendIdToIdMap.get(backendTodo.columnId) ?? backendTodo.columnId;
      if (!todosMap[columnId]) {
        todosMap[columnId] = [];
      }
      todosMap[columnId].push({
        ...backendTodo,
        columnId,
        id: uuid(),
        backendId: backendTodo.id,
      });
    });

    set({ todos: todosMap, columns: columnMap, user: initialUser });
  },

  addTodo: (newTodo) => {
    const result = { ...newTodo, id: uuid() };
    const todos = produce(get().todos, (draft) => {
      if (!draft[newTodo.columnId]) {
        draft[newTodo.columnId] = [];
      }

      draft[newTodo.columnId].push(result);
    });

    set({ todos });
    return result;
  },

  moveTodo: (newTodo, fromColumnId, toColumnId, position) => {
    const todos = produce(get().todos, (draftState) => {
      const sourceColumn = (draftState[fromColumnId] ?? []).filter(
        (todo) => todo.id !== newTodo.id,
      );

      let destinationColumn = draftState[toColumnId] ?? [];
      if (fromColumnId === toColumnId) {
        destinationColumn = sourceColumn;
      }

      if (position <= destinationColumn.length) {
        destinationColumn.splice(position, 0, newTodo);
      } else {
        destinationColumn.push(newTodo);
      }

      draftState[fromColumnId] = sourceColumn;
      draftState[toColumnId] = destinationColumn;
    });

    set({ todos });
  },

  addColumn: (newColumn) => {
    const currentState = get();
    const newId = uuid();
    const todos = produce(currentState.todos, (draft) => {
      draft[newId] = [];
    });
    const columns = produce(currentState.columns, (draft) => {
      const position = Object.keys(currentState.columns).length;
      draft[newId] = { ...newColumn, id: newId, position };
    });

    set({ columns, todos });
    return columns[newId];
  },
});

export const useZustand = create(stateCreator);
