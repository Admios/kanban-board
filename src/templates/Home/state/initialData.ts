import { Column, Todo } from "./types";
import { v4 as uuid } from "uuid";

export const columnNew: Column = {
  name: "New",
  position: 0,
  color: "black",
  id: "columnNew",
};

export const columnNewTodos: Todo[] = [
  {
    id: uuid(),
    text: "Task 17",
  },
  {
    id: uuid(),
    text: "Task 28",
  },
  {
    id: uuid(),
    text: "Task 39",
  },
];

export const columnInProgress: Column = {
  name: "In Progress",
  position: 1,
  color: "red",
  id: "columnInProgress",
};

export const columnInProgressTodos: Todo[] = [
  {
    id: uuid(),
    text: "Task 45",
  },
  {
    id: uuid(),
    text: "Task 56",
  },
];

export const columnInReview: Column = {
  name: "Review",
  position: 2,
  color: "green",
  id: "columnReview",
};

export const columnInReviewTodos: Todo[] = [
  {
    id: uuid(),
    text: "Task 87",
  },
  {
    id: uuid(),
    text: "Task 29",
  },
  {
    id: uuid(),
    text: "Task 63",
  },
  {
    id: uuid(),
    text: "Task 4",
  },
];

export const columnDone: Column = {
  name: "Done",
  position: 3,
  color: "blue",
  id: "columnDone",
};

export const columnDoneTodos: Todo[] = [
  {
    id: uuid(),
    text: "Task 7",
  },
  {
    id: uuid(),
    text: "Task 8",
  },
  {
    id: uuid(),
    text: "Task 9",
  },
];
