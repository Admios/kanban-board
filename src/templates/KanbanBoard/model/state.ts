import { StateDTO } from "@/model/State";
import { TaskDTO } from "@/model/Task";
import { UserDTO } from "@/model/User";
import { Immutable } from "immer";
import { State, Task } from "./types";

export type KanbanState = Immutable<{
  /**
   * The order of the states in the board.
   */
  statesOrder: string[];

  /**
   * The order of the tasks in each state.
   */
  tasksOrder: Record<string, string[]>;

  /**
   * Key: taskId. Value: Task.
   */
  tasks: Record<string, Task>;

  /**
   * Key: stateId. Value: State.
   */
  states: Record<string, State>;

  /**
   * The user that is currently logged in
   */
  user?: UserDTO;
}>;

export interface KanbanActions {
  setUser(user?: UserDTO): void;
  initialize(initialTasks: TaskDTO[], initialStates: StateDTO[]): void;
  addTask(newTask: Task): Task;
  moveTask(
    newTask: Task,
    fromStateId: string,
    toStateId: string,
    position: number,
  ): void;
  addState(newState: State): State;
  editTask(id: string, updatedValues: Partial<Task>): void;
  deleteTask: (id: string) => void;

  clear(): void;
}
