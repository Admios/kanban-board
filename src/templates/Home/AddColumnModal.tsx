import {
  Button,
  FormControl,
  FormErrorMessage,
  FormHelperText,
  FormLabel,
  Input,
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalFooter,
  ModalHeader,
  ModalOverlay,
} from "@chakra-ui/react";
import { KeyboardEventHandler, useState } from "react";
import { useZustand } from "./state";
import { addColumnDB } from "./homeServerActions";
import { v4 as uuid } from "uuid";

interface AddModalProps {
  isOpen: boolean;
  onClose: () => void;
}

export function AddColumnModal({ isOpen, onClose }: AddModalProps) {
  const columns = useZustand((store) => store.columns);
  const addColumn = useZustand((store) => store.addColumn);
  const user = useZustand((store) => store.user);
  const [title, setTitle] = useState("");
  const isError = title === "";

  const submit = () => {
    if (!user) { return }
    const newColumn = { name: title, id: uuid(), color: "black", position: Object.values(columns).length, owner: user.username }
    addColumnDB(newColumn);
    addColumn(newColumn);
    handleClose();
  };

  const handleClose = () => {
    onClose();
    setTitle("");
  };

  const submitOnEnter: KeyboardEventHandler<HTMLInputElement> = (event) => {
    if (event.key === "Enter" && !isError) {
      submit();
    }
  }

  return (
    <Modal isOpen={isOpen} onClose={onClose} isCentered={true}>
      <ModalOverlay />
      <ModalContent>
        <ModalHeader>Add Column</ModalHeader>
        <ModalCloseButton />
        <ModalBody>
          <FormControl isInvalid={isError}>
            <FormLabel>Name:</FormLabel>
            <Input
              type="text"
              value={title}
              onChange={(event) => setTitle(event.target.value)}
              autoFocus
              onKeyUp={submitOnEnter}
            />

            {!isError ? (
              <FormHelperText>The name of the new column.</FormHelperText>
            ) : (
              <FormErrorMessage>Name is required.</FormErrorMessage>
            )}
          </FormControl>
        </ModalBody>
        <ModalFooter>
          <Button colorScheme="yellow" mr={3} onClick={handleClose}>
            Close
          </Button>
          <Button
            colorScheme="blue"
            mr={3}
            onClick={submit}
            isDisabled={isError}
          >
            Add Column
          </Button>
        </ModalFooter>
      </ModalContent>
    </Modal>
  );
}
