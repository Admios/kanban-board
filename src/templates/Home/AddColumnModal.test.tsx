import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { AddColumnModal } from "./AddColumnModal";
import { useZustand } from "./state";


function setupDialog() {
  useZustand.setState({
    columns: {},
    todos: {},
  });

  const onClose = jest.fn();
  render(<AddColumnModal isOpen={true} onClose={onClose} />);
  return { onClose };
}

function tearDownDialog() {
  useZustand.setState({
    columns: {},
    todos: {},
  });
}

it("should add a column when pressed", async () => {
  const { onClose } = setupDialog();

  await userEvent.type(screen.getByRole("textbox"), "My new column");
  await waitFor(() => {
    expect(screen.getByLabelText("Close").getAttribute("disabled")).toBeNull();
  });
  await userEvent.click(screen.getByText("Add Column", { selector: "button" }));

  expect(onClose).toHaveBeenCalled();
  const columns = useZustand.getState().columns;
  expect(Object.values(columns)).toHaveLength(1);
  const columnId = Object.keys(columns)[0];
  expect(columns[columnId]).toMatchSnapshot(
    {
      id: expect.any(String),
    },
    "Columns Result",
  );

  tearDownDialog();
});

it("should focus on the input component and submit when 'Enter' is pressed", async () => {
  const { onClose } = setupDialog();
  const textbox = screen.getByRole("textbox");

  expect(textbox).toHaveFocus();
  await userEvent.type(textbox, "My new column 2{enter}");
  expect(onClose).toHaveBeenCalled();
  const columns = useZustand.getState().columns;
  expect(Object.values(columns)).toHaveLength(1);
  const columnId = Object.keys(columns)[0];
  expect(columns[columnId]).toMatchSnapshot(
    {
      id: expect.any(String),
    },
    "Columns Result",
  );

  tearDownDialog();
})

