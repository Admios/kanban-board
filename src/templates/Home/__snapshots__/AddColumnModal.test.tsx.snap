// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add a column when pressed: Columns Result 1`] = `
{
  "backendId": null,
  "color": "black",
  "id": Any<String>,
  "name": "My new column",
  "position": 0,
}
`;

exports[`should focus on the input component and submit when 'Enter' is pressed: Columns Result 1`] = `
{
  "backendId": null,
  "color": "black",
  "id": Any<String>,
  "name": "My new column 2",
  "position": 0,
}
`;
